signature/src/signer.rs

```diff
@@ -0,0 +1,58 @@
+use crate::error::SignatureError;
+use crate::suite::{Ed25519Signature, Signature, VerificationRelation};
+
+pub trait DIDSigner<S>
+where
+    S: Signature,
+{
+    fn sign(&self, msg: &[u8]) -> S {
+        self.try_sign(msg).expect("signature operation failed")
+    }
+
+    fn try_sign(&self, msg: &[u8]) -> Result<S, SignatureError>;
+    fn get_proof_type(&self) -> String;
+    fn get_verification_method(&self, relation: VerificationRelation) -> String;
+    fn encoded_sign(&self, data: &[u8]) -> String {
+        let signature = self.sign(data);
+        return self.encode(signature);
+    }
+    fn try_encoded_sign(&self, data: &[u8]) -> Result<String, SignatureError> {
+        let signature = self.try_sign(data)?;
+        return Ok(self.encode(signature));
+    }
+    fn encode(&self, sig: S) -> String;
+}
+
+pub struct Ed25519DidSigner {
+    private_key: ed25519_zebra::SigningKey,
+    pub(crate) public_key: ed25519_zebra::VerificationKey,
+}
+
+impl Ed25519DidSigner {
+    pub fn new() -> Self {
+        let sk = ed25519_zebra::SigningKey::new(rand::thread_rng());
+
+        return Self {
+            private_key: sk,
+            public_key: ed25519_zebra::VerificationKey::from(&sk),
+        };
+    }
+}
+
+impl DIDSigner<Ed25519Signature> for Ed25519DidSigner {
+    fn try_sign(&self, data: &[u8]) -> Result<Ed25519Signature, SignatureError> {
```

Comment:
In some cases we have newline between methods, in some we don't, maybe we have to be consistent here